### PR TITLE
NNX migration prep (1/N): pure_nnx flag and init_state_fn scaffolding

### PR DESCRIPTION
**Pull request by ecnal-cienet**
# NNX Migration Route Map
1. :white_check_mark: Add NNX support function and utils / pure_nnx flag=False. Won't affect current Linen workflow.
2. :x: NNX fully supported. pure_nnx flag=False, but user can do NNX runs/tests.
3. :x: NNX unit tests and performance tests are completed and verified. pure_nnx flag=True.
4. :x: Remove Linen code related NNX flags.

# Description

> **Note:** This is the first in a series of NNX migration PRs. Pure NNX training is **not yet implemented** — all NNX code paths currently raise `NotImplementedError`. This PR only introduces the structural scaffolding needed for subsequent patches to plug in NNX logic without modifying shared infrastructure.

This PR introduces two abstractions that enable incremental NNX migration while keeping existing Linen code fully functional.

## `pure_nnx` config flag
A boolean (`configs/base.yml`, `configs/types.py`) that will route all major code paths — training, compilation, inference, RL, and utilities — to pure-NNX logic when `True`, falling back to Linen otherwise. Defaults to `False` so all existing behaviour is unchanged.

## `init_state_fn`
A pluggable callable for initializing the model training state, threaded through `create_checkpoint_manager`, `setup_training_state`, `setup_decode_state`, and `get_abstract_state`. This decouples state
initialization from shared infrastructure so future NNX and Linen paths can provide their own implementations without forking utilities.

## Other structural changes
- `create_training_tools` is split into `create_training_optimizer` and `create_checkpoint_manager` for cleaner separation of concerns.
- `jit_train_step` gains a `mesh` parameter to accommodate NNX callers where `model` is a `GraphDef` with no `.mesh` attribute.
- `get_shaped_inputs` in `train_compile.py` adds a `pure_nnx` branch that omits `example_rng`, matching the future NNX `train_step(state, batch)` signature.
- `get_first_step` restored to the two-argument `(model, state)` form to support both Linen `TrainState` and NNX `TrainStateNNX` step retrieval.
- All entry points (`grpo_trainer`, `maxengine`, `generate_param_only_checkpoint`, `layerwise_quantization`, `lora_utils`, `standalone_checkpointer`, integration tests) updated to accept and pass `init_state_fn`.

## New files
- `src/maxtext/layers/train_state_nnx.py` — NNX `TrainStateNNX` container wrapping `nnx.Module` + `nnx.Optimizer` (mirrors Linen `TrainState`).
- `src/maxtext/utils/maxtext_utils_nnx.py` — NNX-specific utilities: abstract state, named sharding helpers, and sharded model creation.

## Lint / test fixes
- `maxtext_utils.py` — remove unused `ShardMode` import.
- `maxtext_utils_test.py` — fix duplicate `Any` import; restore `Transformer` alias; update `get_abstract_state` call to new `(config, mesh, init_state_fn)` signature.
- `sharding_compare_test.py` — add `pure_nnx=False`/`enable_nnx=False`/ `pure_nnx_decoder=False` to config params; update `get_abstract_state` and `get_logical_annotations` calls to new API.
- `state_dtypes_test.py` — update `get_abstract_state` call to new API.

# Tests
```bash
python3 -m pytest tests/unit/train_utils_test.py -v
python3 -m pytest tests/unit/train_compile_test.py -v
python3 -m pytest tests/unit/maxtext_utils_test.py -v
python3 -m pytest tests/unit/state_dtypes_test.py -v
python3 -m pytest tests/unit/sharding_compare_test.py -v
```

# Checklist
Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/train_state_nnx.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    Mimics the Linen apply_gradients function.
+    Updates the optimizer state, applies updates to parameters,
+    and increments the step counter.
+    """
+    if self.optimizer is None:
+      raise RuntimeError(
+          "Cannot call apply_gradients on a TrainStateNNX initialized without an optimizer. "
+          "This usually happens when the state was created for inference only."
+      )
+    self.optimizer.update(self.model, grads)
```
Comment (bvandermoon) [inline src/maxtext/layers/train_state_nnx.py:48]: @cgarciae is there a way to make this work through bridging?

File: src/maxtext/utils/maxtext_utils_nnx.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+from maxtext.utils import max_logging
+from maxtext.configs import pyconfig
+
+
+def create_nnx_rngs(
+    config: pyconfig.HyperParameters, is_training: bool = True, rng_key: jax.Array | None = None
+) -> nnx.Rngs:
+  """
+  Create NNX Rngs
```
Comment (bvandermoon) [inline src/maxtext/utils/maxtext_utils_nnx.py:31]: Is a better name possible for this file? I think this name will no longer work well once everything is NNX

File: src/maxtext/trainers/pre_train/train_compile.py [hunk 3]
```diff
@@ -93,7 +97,10 @@ def get_shaped_inputs(topology_mesh, config):
   """Get shaped abstractions of inputs to train_step: state, batch and rng"""
   # Construct the model and optimizer to get shaped versions of the state
   quant = quantizations.configure_quantization(config)
-  model = Transformer(config, topology_mesh, quant=quant, model_mode=MODEL_MODE_TRAIN)
+  if config.pure_nnx:
+    _create_model_partial, model = model_creation_utils.create_nnx_abstract_model(config, topology_mesh)
```
Comment (bvandermoon) [inline src/maxtext/trainers/pre_train/train_compile.py:101]: I am not able to find this `create_nnx_abstract_model` for some reason, mind pointing me to where it's located?

File: src/maxtext/checkpoint_conversion/standalone_scripts/convert_gpt3_ckpt_from_paxml.py [hunk 4]
```diff
@@ -87,7 +88,10 @@ def convert(paxml_ckpt_path, maxtext_model_name, base_output_directory, run_name
   mesh = Mesh(devices_array, cfg.mesh_axes)
 
   quant = quantizations.configure_quantization(cfg)
-  model = transformer_as_linen(cfg, mesh, quant=quant, model_mode=MODEL_MODE_TRAIN)
+  if cfg.pure_nnx:
+    raise NotImplementedError("Pure NNX support has not been implemented yet.")
```
Comment (bvandermoon) [inline src/maxtext/checkpoint_conversion/standalone_scripts/convert_gpt3_ckpt_from_paxml.py:92]: The plan is to point to a bridged Transformer version soon, right? Not a forked version

File: src/maxtext/layers/train_state_nnx.py (hunk 1 above)
Comment (cgarciae) [reply src/maxtext/layers/train_state_nnx.py:48]: You could bridge `TrainStateNNX` but I don't think this is what we want here.

What is the issue here?

Comment (ecnal-cienet) [reply src/maxtext/layers/train_state_nnx.py:48]: Agreed with Cristian. Bridging TrainStateNNX would export it as a Linen module, which loses the native NNX semantics we rely on for nnx.split/nnx.merge, checkpointing, and sharding.

File: src/maxtext/utils/maxtext_utils_nnx.py (hunk 2 above)
Comment (ecnal-cienet) [reply src/maxtext/utils/maxtext_utils_nnx.py:31]: The _nnx suffix is intentional during the transition period to clearly separate NNX-specific utilities from the Linen ones in maxtext_utils.py. Once the migration is complete, the plan is to rename maxtext_utils_nnx.py → maxtext_utils.py (replacing the old one) or merge the NNX utilities directly in. But I am also willing to rename to something like nnx_utils.py now if you preferred.

File: src/maxtext/trainers/pre_train/train_compile.py (hunk 3 above)
Comment (ecnal-cienet) [reply src/maxtext/trainers/pre_train/train_compile.py:101]: Fixed in the latest commit: added create_nnx_abstract_model() which runs nnx.eval_shape to build the abstract model for AOT compilation without sharding or checkpointing. Also fixed a related bug where shaped_train_args was missing the dropout_rng slot (set to None for NNX since the train step always takes that positional arg).

File: src/maxtext/checkpoint_conversion/standalone_scripts/convert_gpt3_ckpt_from_paxml.py (hunk 4 above)
Comment (ecnal-cienet) [reply src/maxtext/checkpoint_conversion/standalone_scripts/convert_gpt3_ckpt_from_paxml.py:92]: Yes, exactly. The NotImplementedError is a deliberate placeholder — NNX support for this script will use the bridged Transformer (not a fork) once that's ready. It will be addressed in later PRs. No separate fork intended.
